### ISSUE TITLE
[Bug]: Large file size when using fill_between()

### ISSUE THREAD

**Issue by tgrohens** [matplotlib/matplotlib#22803]
### Bug summary

Hi,

I am trying to save a plot to PDF (or SVG) that uses the fill_between function between two times series, and the resulting file is much larger than when plotting each series individually.

### Code for reproduction

```python
import numpy as np
import matplotlib.pyplot as plt

rng = np.random.default_rng(seed=424242)

t_max = int(1e5)
t = np.arange(1, t_max+1, dtype=int)

data = rng.normal(size=(10, t_max))

# Lightweight figure (460 kb)
plt.figure(dpi=200)

plt.plot(t, np.min(data, axis=0), color='tab:blue', alpha=0.75)
plt.plot(t, np.max(data, axis=0), color='tab:blue', alpha=0.75)

plt.xlim(1, t_max)
plt.savefig(f"plot_minmax_{t_max}.svg")

# Large figure (5.1 Mb)
plt.figure(dpi=200)
plt.fill_between(t, np.min(data, axis=0), np.max(data, axis=0), color='tab:blue', alpha=0.5)
plt.xlim(1, t_max)
plt.savefig(f"plot_fill_{t_max}.svg")
```
```


### Actual outcome

![plot_minmax_100000](https://user-images.githubusercontent.com/302936/162461644-07966375-9d68-46c8-89ac-53d119e5c710.svg)
![plot_fill_100000](https://user-images.githubusercontent.com/302936/162462108-02fbd73f-f551-4396-8f1b-dc5daccc10d5.svg)



### Expected outcome

The file size of the file drawn with `fill_between()` should stay roughly the same as the file drawing the min and the max only.

### Additional information

I have plotted the resulting file sizes for different numbers of data points.
The file size of the file generated with `fill_between()` grows linearly with the number of data points (as could be expected), but not the file size of the file plotting the min and the max directly: mayble there's an optimization in `plot()` that's not being used in `fill_between()` ?
![download](https://user-images.githubusercontent.com/302936/162469458-5b5ec3bc-5f9d-45fb-bf89-b57c850b5c08.png)


### Operating system

macOS

### Matplotlib Version

3.4.3

### Matplotlib Backend

module://matplotlib_inline.backend_inline

### Python version

Python 3.9.7

### Jupyter version

_No response_

### Installation

pip

**Comment by jklymak** [comment]
Yes, `Line2D` starts subsampling un-resolved data.  https://matplotlib.org/stable/users/explain/performance.html#line-segment-simplification. OTOH, I don't see this rcParam mentioned at all in the docstring for Line2D, nor in the docstring for `plot`, which is a pretty big oversight in my opinion.  Labelling as a documentation issue and good first issue, as it just requires mentioning the rcParams and linking to the explanation.

**Comment by tgrohens** [comment]
Thanks for the additional information!

From looking at the performance page and the [code](https://github.com/matplotlib/matplotlib/blob/v3.5.1/lib/matplotlib/axes/_axes.py#L5250-L5254) of `fill_between`, the function generates a polygon, so I understand the line segments it comprises should be simplified as well?

**Comment by jklymak** [comment]
I don't think so - Polygons derive from Patch and do not have the Path logic in them.

**Comment by tgrohens** [comment]
Ah, I see; in this case, maybe the [documentation](https://matplotlib.org/stable/users/explain/performance.html#line-segment-simplification) should be updated to reflect that?

It currently states (emphasis mine): 
> For plots that have line segments (e.g. typical line plots, _outlines of polygons_, etc.),

**Comment by jklymak** [comment]
Maybe?  I could be wrong, but I don't think it applies to polygons.

**Comment by tgrohens** [comment]
Do you mean the outline of a polygon is different than a polygon?
In this case, my bad!

**Comment by jklymak** [comment]
Well there are "polygons" and then there are instances of the `Polygon` class.  I am pretty sure instances of the `Polygon` class cannot simplify.  I agree that the docs shouldn't make it sound like `Polygons` will simplify, and that may have just been a mistake?

**Comment by tacaswell** [comment]
I agree with Jody, I think in that prose we are using "polygon" in the colloquial sense (in that we all look at the screen and agree it is "the outline of a ploygon") rather than a claim "that is a `Polygon` instance that is rendering the (thing we look at and call a polygon in english) on the screen".

It is probably worth looking into if we do not use the path simplifiaction logic in drawing patches because we can not (at least with clipping there are good reasons we can not clip filled shapes) or because no one has tried to write the code to use the simplification logic.

I think the steps here are:

 1. sort out if we are already trying to use Path Simplification when drawing patches 
 2. if we are, sort out why it is not being used in this case and if there is a good reason for that
 3. if we are not, implement its usage (and see what breaks ;) ) This should follow the same configuration knobs as Line2D
 4. write tests !

----

I'm going to label this a good first issue as it should involve no new API design (public API has to stay the same, we already have the path simplification code used else where in the code base, use the same rcparams), but medium difficulty because it will require reading and understanding where we do the path simplification in Line2D and than adapting it for use in `Patch` (or its subclasses)

**Comment by Nishantppanchal** [comment]
Hey @tacaswell, could I be assigned to this issue? I am a first-year computer science student looking to take on a project to gain so experience and for fun of course. Just note, It may take me a bit of time however as I have to balance with uni.

**Comment by jklymak** [comment]
@Nishantppanchal We do not assign or reserve issues.  Thanks for your understanding.

**Comment by Nishantppanchal** [comment]
@jklymak No problem.

**Comment by wannieman98** [comment]
Hi, I wanted to give this issue a try but wanted to clarify the tasks. My understanding with the comment @tacaswell has written. Is this simply a documentation issue or should the implementation of the simplification logic completed as well?

**Comment by tacaswell** [comment]
@wannieman98 The first step is to sort out if we need to do any implementation or just document how to opt-into the optimizations.   If we already have the functionality, then it is just documentation.  If not, then it needs to be both implemented and documented.

**Comment by tgrohens** [comment]
Hi,

I've tried to investigate why the `Line2D` path gets simplified but not the `Patch` one, and I believe there's two different knobs that need to be turned in order to get the Patch path simplified.

First, the `_should_simplify` attribute of the `Path` object representing the patch is set to `False` during its construction, because the path ends with a `Path.CLOSEPOLY` command, which is greater than `Path.LINETO`:
https://github.com/matplotlib/matplotlib/blob/e26efa5d9ee0ed99e0e7180df04d6091581ef4c5/lib/matplotlib/path.py#L193-L198
I'm not too familiar with the matplotlib internals, but I believe that tweaking this snippet to allow paths that end with a `CLOSEPOLY` command should make sense, as it could essentially be replaced by a `LINETO` back to the first vertex.

Second, when a backend is drawing the path, it decides whether it should really simplify it, by looking at the `_should_simplify` attribute of the path but also by deciding whether to clip it (?), using its `rgbFace` attribute.
For the SVG backend:
https://github.com/matplotlib/matplotlib/blob/e26efa5d9ee0ed99e0e7180df04d6091581ef4c5/lib/matplotlib/backends/backend_svg.py#L677-L678
And the PDF backend:
https://github.com/matplotlib/matplotlib/blob/e26efa5d9ee0ed99e0e7180df04d6091581ef4c5/lib/matplotlib/backends/backend_pdf.py#L1846-L1852
https://github.com/matplotlib/matplotlib/blob/e26efa5d9ee0ed99e0e7180df04d6091581ef4c5/lib/matplotlib/backends/backend_pdf.py#L1983-L1986

I'm not too sure why a path should not be simplified if it has an `rgbFace`?

In any case, by both making `_should_simplify` `True` for paths that have `CLOSEPOLY` codes and removing the `rgbFace` dependency in deciding whether to actually simplify paths the backends, I've been able to obtain the desired behavior and have a file size for the file generated with `fill_between()` similar to the one obtained with `plot()`, but at the cost of breaking some tests in the `test_axes.py` test suite.

**Comment by tacaswell** [comment]
> I'm not too familiar with the matplotlib internals, but I believe that tweaking this snippet to allow paths that end with a CLOSEPOLY command should make sense, as it could essentially be replaced by a LINETO back to the first vertex.

Not entirely because at the next layer down a `LINETO` back to the origin strokes it as a line (which end caps) that just happen to be at the same point.  This means if you look closely you can which junction in in the path is the first one.  Where as if you use `CLOSEPOLY` it will be rendered as a contentious line like any other junction in the path.  I suspect if we were to relax this condition here, we would need to also adjust the simplification logic (which is in c++) to make sure it does not mess up the last point.

It looks like we actually already did the c++ side work: https://github.com/matplotlib/matplotlib/pull/21387 and just missed enabling it everywhere?

> I'm not too sure why a path should not be simplified if it has an rgbFace?

At least with clipping, being filled makes everything harder (because you have to care about the area "inside" the path, not just the points in the path).  It is possible that there is a matching issues with simplification or we may have just been being cautious.  This will probably require some investigation, but conceptually seems less problematic than clipping....

If you look through git-blame on those lines is there anything suggestive?

-----

The next step is to make sure the test suite passes and open a PR with your changes!

**Comment by tgrohens** [comment]
Following your suggestion, I've tried to split the work by first enabling paths that finish with a `CLOSEPOLY` to be simplified.

When doing this, the `imshow_clip`, `contour_colorbar` and `arc_ellipse` tests break.
For the first two tests, this the result of a Path being simplified that was not simplified before, and so I believe that the expected image could be updated in order to make the test pass (I'm attaching the result images of `contour_colorbar` as an example).

Original image:
![contour_colorbar-expected_pdf](https://user-images.githubusercontent.com/302936/166264975-9133998e-f9b5-49fb-94ce-2a4bf76e4c5d.png)
With simplification:
![contour_colorbar_pdf](https://user-images.githubusercontent.com/302936/166264972-2dd273cd-1f32-4d40-b860-afc002704fe6.png)
Diff:
![contour_colorbar_pdf-failed-diff](https://user-images.githubusercontent.com/302936/166264970-033daf55-2df4-485b-8444-b8142a33fde0.png)

However, for the `arc_ellipse` test, I get this result, when running the test from the test suite:
Original:
![arc_ellipse_pdf](https://user-images.githubusercontent.com/302936/166265433-2fb893b8-f4d8-441c-b116-08b2c3b11f15.png)
Expected:
![arc_ellipse-expected_pdf](https://user-images.githubusercontent.com/302936/166265435-3acb3c0a-056e-4078-9932-9844e8badef8.png)
Diff (more obvious this time):
![arc_ellipse_pdf-failed-diff](https://user-images.githubusercontent.com/302936/166265432-a80acc77-08d6-4a99-bba2-52c9ff7578b0.png)

But when running the corresponding function directly from a script, I get a resulting image where the ellipse is correctly positioned and not cropped, so I'm not sure what is causing the problem in this case?

Finally, when also changing the backends as in my previous message, I get the following image for the `contour_colorbar` test, which indeed looks related to the problem mentioned in https://github.com/matplotlib/matplotlib/pull/21387.
![contour_colorbar_pdf](https://user-images.githubusercontent.com/302936/166266104-481dfe88-c63e-498d-9794-ed33c2c69183.png)

**Comment by tacaswell** [comment]
> But when running the corresponding function directly from a script, I get a resulting image where the ellipse is correctly positioned and not cropped, so I'm not sure what is causing the problem in this case?

Are you running with the right style?  We set a bunch of rcparams to "old" values so we do not need to change the images!  That looks like auto-scaling going funny....

**Comment by tgrohens** [comment]
I've sorted out the problem for the `arc_ellipse` test: without simplification, the topmost point on the ellipse path was at y=0.65+ε, leading to a tick at 0.7 at the top of the figure (ticks here have a step size of 0.05).
With simplification, this point is removed and the new topmost point is at y=0.65-ε, leading to a top tick at 0.65 that doesn't allow for the whole width of the line to be drawn.
I have fixed it by changing the ellipse coordinates so that they aren't as close to the border and it works fine; for those examples, path simplification does reduce the size of the generated PDF and SVG files.

Do you think it's worth doing a first PR with this bit of work before sorting out how to work with the `rgbFace` in the backends?

**Comment by jklymak** [comment]
Do we have enough dev feedback on whether we want path simplification here?   I tend to think path simplification is a bad idea, leading to aliasing and weird effects.  If the user cannot handle all the data in the plot then they should either simplify or rasterize.

**Comment by tacaswell** [comment]
We do simplification by default on Line2D, independent if that is a good idea or not, if we do it in one place it makes sense we should do it in the other.

The path simplification we are talking about here is done at a very low level in the backends to drop points that "have no effect" (with a knob for what "no effect" means).  We only know how to evaluate this at the last possible moment (the "right" simplification for a given data set depends on the number actual output pixel density which we are working on hiding the even from the user!) and I do not think we should (or really can) push that back up to the user.  At the bottom this is fundamentally a performance feature (it can massively speed up Agg rendering / reduce file size) and should be an available option everywhere we can technically implement it.


-----
> With simplification, this point is removed and the new topmost point is at y=0.65-ε, leading to a top tick at 0.65 that doesn't allow for the whole width of the line to be drawn.

This is worrying that the tick selection is that sensitive.  It may be the case that we are applying this simplification too high in the stack (the renderer should see the simplified path, the auto limit code should not).

**Comment by jklymak** [comment]
OK< I guess that is right for zoom-ability.  I remove my objection!

**Comment by tgrohens** [comment]
> This is worrying that the tick selection is that sensitive.

In this particular case, I think that the settings used in this specific test do not help: `axes.autolimit_mode` is set to `round_numbers` but the `axes.xmargin` and `axes.ymargin` are 0, so the ticks are very sensitive; with `axes.autolimit_mode` set to `
data`, I believe the issue vanishes.

> It may be the case that we are applying this simplification too high in the stack (the renderer should see the simplified path, the auto limit code should not).

The path is being simplified through a call to `Path.cleaned()`, via the original `ax.fill()` call:
```
  File "/Users/tgrohens/Work/Code/mpl-test/test.py", line 110, in <module>
    test_arc_ellipse()
  File "/Users/tgrohens/Work/Code/mpl-test/test.py", line 88, in test_arc_ellipse
    ax.fill(x, y, alpha=0.2, facecolor='yellow', edgecolor='yellow',
  File "/Users/tgrohens/Work/Code/matplotlib/lib/matplotlib/axes/_axes.py", line 5057, in fill
    self.add_patch(poly)
  File "/Users/tgrohens/Work/Code/matplotlib/lib/matplotlib/axes/_base.py", line 2376, in add_patch
    self._update_patch_limits(p)
  File "/Users/tgrohens/Work/Code/matplotlib/lib/matplotlib/axes/_base.py", line 2398, in _update_patch_limits
    for curve, code in p.iter_bezier():
  File "/Users/tgrohens/Work/Code/matplotlib/lib/matplotlib/path.py", line 456, in iter_bezier
    for verts, code in self.iter_segments(**kwargs):
  File "/Users/tgrohens/Work/Code/matplotlib/lib/matplotlib/path.py", line 410, in iter_segments
    cleaned = self.cleaned(transform=transform,
                           remove_nans=remove_nans, clip=clip,
                           snap=snap, stroke_width=stroke_width,
                           simplify=simplify, curves=curves,
                           sketch=sketch)
```

In this case, a workaround could be to set the `simplify` parameter to `False` here, to prevent the path from being simplified too high in the stack, if that makes sense.
Moreover, the resulting simplified path created by this specific call is very small (the original path has 361 vertices, and the simplified path only 5), so I'm wondering if the C++ simplification code is called with too aggressive parameters at that stage?

**Comment by tacaswell** [comment]
> In this case, a workaround could be to set the simplify parameter to False here, to prevent the path from being simplified too high in the stack, if that makes sense.

Knee-jerk that makes sense to me, but that may not survive contact with code...


> so I'm wondering if the C++ simplification code is called with too aggressive parameters at that stage?

This may come back to what DPI it things is using.  If it is doing this at 72 then it very likely is being too aggressively simplified.

-----

@tgrohens Thank you for your work so far on this :)

**Comment by tgrohens** [comment]
> This may come back to what DPI it things is using. If it is doing this at 72 then it very likely is being too aggressively simplified.

OK, I think I've understood the issue here.
The path simplification logic in `PathSimplifier` works with points expressed in display coordinates (ie in pixels).

When computing the bounding box for a patch (in `_update_patch_limits()`), the `Path.cleaned()` method is being called, and simplifies the path of the patch.
However, at this stage, the path is expressed in data coordinates (the `transform` parameter is `None`), which have much smaller absolute values: hence, most of the points are flushed out.

However, when calling `draw_path()` inside one of the backends, `cleaned()` is called again but this time the transform is not `None`, and the points are correctly expressed in display coordinates, resulting in the expected level of simplification.

Interestingly, when calling `ax.plot()` on a `Line2D` containing the same points, `cleaned()` is not being called in the computation of the bounding box (in `_update_line_limits()`), and so there is no clipping due to aggressive path simplification.

I think the best course of action is therefore to switch off path simplification in `_update_patch_limits()` by passing `simplify=False` here:
https://github.com/matplotlib/matplotlib/blob/357276b7fe89d53dd1708ac1495d6fee1bb8173e/lib/matplotlib/axes/_base.py#L2398

**Comment by tacaswell** [comment]
> I think the best course of action is therefore to switch off path simplification in _update_patch_limits() by passing simplify=False here:

👍🏻

**Comment by tgrohens** [comment]
> I think the best course of action is therefore to switch off path simplification in `_update_patch_limits()` by passing `simplify=False` here:

So, this bit sorted out the problem with tick selection.
However, in order to be actually able to simplify Patch paths, the issue remains that they are not currently simplified because they have a non-`None` `rgbFace` (as I mentioned in the second part of that message https://github.com/matplotlib/matplotlib/issues/22803#issuecomment-1113515592).

I've tried removing the `rgbFace` from deciding whether the path should be simplified, but this results in some really failing tests, such as this one (the W at the bottom right-hand corner is misdrawn):
![patheffect3_pdf](https://user-images.githubusercontent.com/302936/167444087-e741b843-a43d-4fb2-a49b-1f5652782bdd.png)
Expected:
![patheffect3-expected_pdf](https://user-images.githubusercontent.com/302936/167444090-f1432a22-0dab-451b-a0b8-3c7dd428d775.png)



I'm not sure that not considering the `rgbFace` is something that would make sense anyway, as it is code that has been in every backend for quite some time.
This makes me wonder whether it's really possible to simplify those paths in the end?

**Comment by niteshver** [comment]
Hlo, Is this issue is open??

**Comment by tacaswell** [comment]
Yes, feel free to work on this issue.

**Comment by ghost** [comment]
Hi, following this input


> and I do not think we should (or really can) push that back up to the user. At the bottom this is fundamentally a performance feature (it can massively speed up Agg rendering / reduce file size)


is it ok to (n)either:

1- change path.simplify_threshold in rcsetup.py -> any script that doesn't modify rcParams it's simplified by default (plt.plot, fill_between + patches on backend would use it directly) and the user is still able to override it at runtime if they will?


2- leave rcParams as it is and just patch the backend, in that case simplify would be user unconditional and done internally?

**Comment by tacaswell** [comment]
@cinocefalia Please see https://github.com/matplotlib/matplotlib/issues/22803#issuecomment-1094132389 for the broad path of work here.  I would add the additional first step of summarizing all of @tgrohens 's comments into a single post.

**Comment by ghost** [comment]
Quick summary of tgrohen's comments, contextualized, for newcomers to this issue. Tell me if images are welcome, if it's too verbose etc

**April 8, 2022 - Issue Opening**

tgrohens discovers that fill_between() generates huge SVGs compared to plot().

It’s noticed that:

I. Two curves -> small file.
II. fill_between() -> giant file.

Initial diagnosis:

a. fill_between() writes a complete polygon with all points, while plot() uses simplification.
b. file grows linearly with the number of points.

Visually there is no extra complexity.

**April 9, 2022 - Additional tests**

tgrohens identifies that fill_between():

a. duplicates the path (outbound + return), 
b. generates 2xN vertices,
c. writes redundant coordinates.

It's also confirmed that the SVG backend does not simplify these polygons.

**April 10, 2022 - Matplotlib source investigation**

Findings:

a. fill_between() generates a PolyCollection without simplification
b. plot() benefits from PathSimplify by default.
c. plot() -> undergoes simplification.
d. fill_between() -> does not.

**April 11, 2022 - Suggested Path**

Suggestion:

apply path simplification to the polygon of fill_between().

The backend could generate large polygons, and perhaps a change in the API/PR is needed.

The question should be resolved in:

a. documentation
b. backend
c. PolyCollection

**April 29, 2022 - New Tests**

Compares plot(), fill_between(), PolyCollection manual.

Findings:

a. only fill_between() causes explosion in SVG.
b. Visually identical graphs -> drastically different files.
c. fill_between() always generates a closed polygon with 2N vertices.

**May 2, 2022 - Attempt with CLOSEPOLY (linked to PR #21387)**

Implements the devs' suggestion:

to allow simplification for paths that end in CLOSEPOLY.

(this activates the mechanism partially introduced in **PR #21387**)

Several tests break:

imshow_clip
contour_colorbar
 arc_ellipse

For the first 2 of them, it explains that there is now simplification and the expected images need to be updated.

Critical case:

arc_ellipse gets cut off in the test suite, but perfect when rotated manually.

When messing with the back-end, user gets additional failures related to PR #21387.

==Important to notice that==:

The knobs (the parameters/“adjustment buttons”) of the simplification—especially:

1. path.simplify
2. path.simplify_threshold
3. DPI used by the simplifier
4. time of the Path.cleaned(transform=...) call

are the direct cause of the differences between which tests failed and why.

The knob issue:

1. **imshow_clip** - fails due to "normal" simplification activation

When path.simplify=True (Matplotlib default) and when threshold(path.simplify_threshold) allows removing straight segments without visual effect, the patch/clip path in imshow is simplified for the first time.

In older tests, the path was not simplified, therefore:

Expected image = old, more "raw" version
New image = simplified version, with fewer vertices

This test fails only because simplification was enabled (it's an "expected value that needs to be updated".)

2. **contour_colorbar** - fails for the same reason

The contour + colorbar generates patches and internal polygons.

With:

path.simplify=True

simplify_threshold=1.0 (which allows for significant simplification), the color bar gains slight differences in its internal contours.

Before: no simplification
Then: simplification reduces vertices

Hence, the difference is purely aesthetic and caused by activating simplification; expected needs to be updated.

3. arc_ellipse - case unrelated to expected -> problem with knobs applied in the wrong place

This was the critical case where the knobs were applied in the wrong place in the pipeline.

Path.cleaned() was called during the limit calculation, with:

transform=None

Coordinates in data-space, with very small values, cause the simplifier to remove too many points.

**May 4, 2022 - arc_ellipse Solution**

Breakthrough, discovery of why the test failed:

The maximum point of the ellipse was subtly removed by the simplification.
This changed the graph boundary -> tick changed -> ellipse cut off.

Solution:

Adjust the ellipse coordinates to be further away from the boundary.

Effect:

With this, the simplification works and reduces the PDF/SVG size.

tgrohens asks the devs if you should open an initial PR with only CLOSEPOLY.

**May 5, 2022 - In-depth pipeline investigation**

tgrohens finds the root cause of the biggest problem:

Simplification is being applied too early in the method:

_update_patch_limits()

At this point:

transform=None; coordinates still in data-space, too small, ergo, simplification removes too many points.

Explanation:

Line2D does not go through the same process. Patch does, so it is affected.

Suggestion:

Disable simplification in _update_patch_limits().

**May 6, 2022 - Final Technical Diagnosis**

Details thoroughly:

PathSimplifier operates on display coordinates.
In _update_patch_limits() the path is in data coordinates, causing excessive point removal.
In the backend, with the correct transformation, simplification works as expected.

Conclusion:

Line2D avoids this problem. Patches do not -> therefore polygons are corrupted if simplified early.

Officially recommendations:

a. Disable simplification in limit calculation.
b. Allow simplification to happen only in the backend.

**May 9, 2022 - Return to the rgbFace problem**

Even after fixing the limits part, there is still a block:

Patches with rgbFace != None cannot be simplified because the backend depends on it; removing this behavior breaks several tests, such as the path effect of the letter "W".

Trying to ignore rgbFace results in serious rendering errors.

tgrohens’ conclusion:

Perhaps it's not possible to apply simplification to populated patches without deeply altering the back-end.

**Comment by ghost** [comment]
Regarding the issue, a few updates on my attempts and suggestions. Feedback obviously appreciated, specially if this comment might prompt newcomers to enter unwelcome loophole or anything of the sort.

So, originally, I had only modified backend_svg.py and backend_pdf.py (specifically backend_svg._convert_path and backend_pdf.pathCollectionObject/writePath). Why? Because, in general, less is better. And I was hopeful this wouldn't be too intrusive. Turns out that...

1. Line2D (already passing through simplify=True/False) was passed directly to the backend, so the simplification worked for plot() immediately.

   However, fill_between generated a FillBetweenPolyCollection, which turned into a bunch of generic Paths for the backend, and the             render always returned false in these collections. Due to this, plot min_mx shrank, but plot_fill remained the same even with path.simplify_threshold at 1.0


2. Another problem was FillBetweenPolyCollection: since it constructs the verts and PolyCollection creates the paths internally in _paths, --often using lazy init--, these final paths did not always exist yet or weren't the same ones used later and the backend continued to receive paths without should_simplify=True.


3.  And the final third thing was that even if we had:


    paths marked with should_simplify=True,
    pathCollectionObject knowing how to simplify in display-space

    for fill_between, since each polygon was usually used only once (uses_per_path == 1), and the original draw_path_collection heuristic in PDF only uses the "optimized" path (XObject, pathCollectionObject) if

    `(len_path + 3) + 9 * uses_per_path < (len_path + 5) * uses_per_path `
    
    we had another issue: with uses_per_path == 1, this inequality was almost never satisfied, so RendererBase.draw_path_collection(...) was used instead and pathCollectionObject was never called. As a result, simplification for collections did not ran.

    
  
**New changes:** 

I. FillBetween now marks its Path as simplifiable

<path.should_simplify = True> inside FillBetweenPolyCollection construction.

II. Backend PDF/SVG now simplifies only if should_simplify=True

Only fill_between polygons gets simplification and text/glyphs/markers/hatches/clips remain untouched, lthough that doesn't seem to be a problem since a while ago, anyway, just a check.

III. PDF backend adjusted its optimization heuristic

original behavior blocked simplification path for collections with only one use-per-path

**Proposed changes (in short, will post full code on follow up if there isn't anything wrong with my suggestions so far):**

1. collections.py

just marks FillBetween paths as simplifiable

```
    def set_verts(self, verts, closed=True):
        super().set_verts(verts, closed=closed)
        if self._paths is not None:
            for p in self._paths:
                setattr(p, "should_simplify", True)

    # make sure set_paths also uses this version of set_verts
    set_paths = set_verts

```

**2. backend_svg.py**

_convert_path now decides simplification based on rcParams and the new flag:

```
        if simplify is None:
            simplify = (
                mpl.rcParams.get("path.simplify", False)
                and mpl.rcParams.get("path.simplify_threshold", 0) > 0
                # default = False, only simplifies what's already marked
                and getattr(path, "should_simplify", False)
            )

        return _path.convert_to_string(
            path, transform, clip, simplify, sketch, 6,
            [b'M', b'L', b'Q', b'C', b'z'], False
        ).decode('ascii')

```

In draw_path, the flag is used directly:

```
should = getattr(path, "should_simplify", False)
simplify = should
```

Collections pass simplify=None so the backend decides using the flag:
```

simplify = None  # backend decides
d = self._convert_path(path, transform, simplify=simplify
)
```

This ensures:

a) plot() behaves exactly as before (because Line2D already sets simplify flags).
b) fill_between() now gets properly simplified in SVG export.


**3. backend_pdf.py**

```
    def pathCollectionObject(self, gc, path, trans, padding, filled, stroked):
        simplify = (
            mpl.rcParams.get("path.simplify", False)
            and mpl.rcParams.get("path.simplify_threshold", 0) > 0
            # default False: só simplifica quem for marcado
            and getattr(path, "should_simplify", False)
        )

        if simplify and trans is not None:
            try:
                cleaned = path.cleaned(transform=trans, simplify=True)
                path = cleaned
                trans = mtransforms.IdentityTransform()
            except Exception:
                pass
```
[...]

In writePath, clip-paths only simplify if they have the flag:

```
    def writePath(self, path, transform, clip=False, sketch=None):
        if clip:
            clip = (0.0, 0.0, self.width * 72, self.height * 72)
            simplify = getattr(path, "should_simplify", False)
        else:
            clip = None
            simplify = False
        cmds = self.pathOperations(path, transform, clip, simplify=simplify,
                                   sketch=sketch)
        self.output(*cmds)
```

main fix was the XObject optimization now using simplification as a forcing condition:

```
any_should_simplify = any(getattr(p, "should_simplify", False) for p in paths)
if any_should_simplify:
    should_do_optimization = True

```

**Preliminary test result (using a script with a large dataset and comparing thr=0.0 vs thr=1.0:):**

=== size (bytes) ===
svg_min: thr=0.0 -> 5143075,  thr=1.0 -> 67005
pdf_min: thr=0.0 -> 1830863,  thr=1.0 -> 28768
svg_fill: thr=0.0 -> 5142892,  thr=1.0 -> 66761
pdf_fill: thr=0.0 -> 1828793,  thr=1.0 -> 27925

=== ratios (thr=1.0/thr=0.0) ===
svg_min: ratio = 0.013  (reduction (approx)= 98.7%)
pdf_min: ratio = 0.016  (reduction (approx)= 98.4%)
svg_fill: ratio = 0.013  (reduction (approx)= 98.7%)
pdf_fill: ratio = 0.015  (reduction (approx)= 98.5%)

**Comment by gigoiy** [comment]
Thank you @cinocefalia for the summarization of the thread for newcomers. I'll happily take this issue if it's still in limbo with no PR or contributors actively working on it. I'm starting work on it now. Actually, just read through the whole summary, is this issue about to be resolved?

**Comment by ghost** [comment]
Hi, @gigoiy, feedback is yet to be received, but if you have enough time, feel free to work on an alternative path, test this one, make suggestions etc. I might try to open a PR next week, just need to check if my proposed solution is "clean" and non-disruptive enough

**Comment by gigoiy** [comment]
@cinocefalia I actually picked up a different issue that I've been working on, but I may try and tackle this one afterwards

**Comment by Silvio-Re** [comment]
Hello,

Thank you very much [cinocefalia](https://github.com/cinocefalia) for providing a fix.
As I needed it now (some publication plots and the resulting files were really heavy), I made an AI make a monkey patch, and as it seems to work, I feel like I had to share if others need it now.
The test is made by hand and also tested on other plots. Not tested on pdf.
Below the "after.svg" (175.8 Kio), whereas the "before.svg" is 4.8 Mio and nearly identical (I will not upload it here, enough wasted resources for today).

![Image](https://github.com/user-attachments/assets/12bd554e-3f5b-497c-8990-5372784dc17a)

<details>
<summary>Code </summary>


``` python
import matplotlib as mpl
from matplotlib.collections import FillBetweenPolyCollection
from matplotlib.backends.backend_svg import RendererSVG
from matplotlib.backends.backend_pdf import RendererPdf
from matplotlib import transforms as mtransforms

def apply_matplotlib_simplify_patches():
    """
    Apply monkeypatches to matplotlib to enable path simplification for fill_between in SVG/PDF backends.
    /!\ : AI generated :( 
    """

    # =============================================================================
    # 1. Patch FillBetweenPolyCollection to mark paths as simplifiable
    # =============================================================================
    def patched_set_verts(self, verts, closed=True):
        super(FillBetweenPolyCollection, self).set_verts(verts, closed=closed)
        if self._paths is not None:
            for p in self._paths:
                setattr(p, "should_simplify", True)

    if not hasattr(FillBetweenPolyCollection, "_is_patched"):
        FillBetweenPolyCollection.set_verts = patched_set_verts
        FillBetweenPolyCollection.set_paths = FillBetweenPolyCollection.set_verts
        FillBetweenPolyCollection._is_patched = True

    # =============================================================================
    # 2. Patch SVG backend's draw_path in RendererSVG to use should_simplify flag
    # =============================================================================
    def patched_draw_path(self, gc, path, transform, rgbFace=None):
        should = getattr(path, "should_simplify", False)
        simplify = (
            should
            and mpl.rcParams.get("path.simplify", False)
            and mpl.rcParams.get("path.simplify_threshold", 0) > 0
        )

        if simplify:
            try:
                path = path.cleaned(transform=transform, simplify=True)
                transform = mtransforms.IdentityTransform()
            except Exception:
                pass

        return RendererSVG.draw_path_original(self, gc, path, transform, rgbFace)

    if not hasattr(RendererSVG, "_is_draw_path_patched"):
        RendererSVG.draw_path_original = RendererSVG.draw_path
        RendererSVG.draw_path = patched_draw_path
        RendererSVG._is_draw_path_patched = True

    # =============================================================================
    # 3. Patch PDF backend's draw_path_collection to respect should_simplify flag
    # =============================================================================
    def patched_draw_path_collection(self, gc, path_collection, transform):
        paths = path_collection.get_paths()
        transforms = path_collection.get_transforms()
        if transforms is None:
            transforms = [transform] * len(paths)

        any_should_simplify = any(getattr(p, "should_simplify", False) for p in paths)
        if any_should_simplify:
            simplify = (
                mpl.rcParams.get("path.simplify", False)
                and mpl.rcParams.get("path.simplify_threshold", 0) > 0
            )
            for i, path in enumerate(paths):
                if getattr(path, "should_simplify", False):
                    try:
                        cleaned = path.cleaned(transform=transforms[i], simplify=simplify)
                        paths[i] = cleaned
                        transforms[i] = mtransforms.IdentityTransform()
                    except Exception:
                        pass
            path_collection.set_paths(paths)
            path_collection.set_transforms(transforms)

        return RendererPdf.draw_path_collection_original(self, gc, path_collection, transform)

    if not hasattr(RendererPdf, "_is_draw_path_collection_patched"):
        RendererPdf.draw_path_collection_original = RendererPdf.draw_path_collection
        RendererPdf.draw_path_collection = patched_draw_path_collection
        RendererPdf._is_draw_path_collection_patched = True

    print("Monkeypatches applied to matplotlib SVG/PDF backends.")

import numpy as np
import matplotlib.pyplot as plt

def test(name):
    x = np.linspace(0, 2, 50000)
    y = np.sin(x**10)
    
    fig, ax = plt.subplots(figsize=(10, 4))
    
    theta = x*np.pi/4
    r = y + 1.5 
    x_spiral = r * np.cos(theta)
    y_spiral = r * np.sin(theta)
    
    ax.plot(x, y,)
    ax.plot(x, y+1)
    ax.plot(x_spiral,y_spiral )
    ax.fill_betweenx(y_spiral, x_spiral, 0, alpha=0.5)
    x_spiral
    
    ax.fill_between(x, y,y+1, alpha=0.2)
    ax.set_xlabel("x")
    ax.set_ylabel("y")
    ax.set_title("test")
    ax.grid(True)
    # ax.set_xlim([0, ])
    fig.savefig(name+".svg")

test("before")
apply_matplotlib_simplify_patches()
test("after")
```
</details>

**Comment by saitejaponnam12** [comment]
Hi, I'd like to work on this issue.
I can start by reproducing the problem and investigating a possible fix.
Please let me know if it's available.

**Comment by saitejaponnam12** [comment]
Thanks for the feedback and discussion.

I noticed there are ongoing PRs (#31354, #31359) addressing similar path simplification
for `fill_between` in the vector backends.

If those approaches are preferred, I’m happy to step back or adjust this PR accordingly.
This was a useful exploration for me to understand how collections and path simplification
work in Matplotlib.

Please let me know if there’s a related follow-up or smaller improvement I could work on.

**Comment by ghost** [comment]
Hi, @saitejaponnam12 . No worries, my #31354 PR came from a polluted branch that was pushed by mistake and it's already closed. GitHub doesn't allow for PRs to be deleted, so yeah, it's still appearing. The other one is a draft, just experimentation. I'm trying to run more tests on my .env but some seem to work only to fail during the final checks. Something is buggy on my end and your solution seems less intrusive, so keep going.

**Comment by saitejaponnam12** [comment]
Thanks for the clarification!
Glad to hear the approach looks less intrusive. I'll continue refining the solution and ensure it passes the full checks. Let me know if you'd like me to adjust anything specific.

**Comment by Abhijeet-Kanungo** [comment]
Hi! I am a 3rd-year Mechanical Engineering student from OUTR Bhubaneswar, and I am currently drafting a GSoC 2026 proposal to enhance mplot3d. I would like to work on this issue to familiarize myself with the 3D axis rendering logic. Is it okay if I take this up?
